Onboarding docs (#3785)
ROCm/rocm-libraries: projects/hipdnn/docs/what-is-hipdnn.rst

.. meta::
   :description: hipDNN is a plugin-based deep learning library that provides graph-based operation support through various backend plugins 
   :keywords: hipDNN, ROCm, library, API

.. _what-is:

******************
What is hipDNN?
******************

hipDNN is a graph-based deep learning library that enables multi-operation fusion for improved performance on AMD GPUs. 
It uses operation graphs as an intermediate representation to describe computations, allowing different backend engines to optimize and execute these graphs efficiently.

hipDNN has a plugin-based architecture which allows you to extend hipDNN without modifying the core library.

> hipDNN allows developers to run deep learning workloads on the frontend on ROCm and AMD GPUs while maintaining compatibility with NVIDIA's cuDNN API.

Reviewer: I think this needs rephrasing:
"hipDNN allows developers to run deep learning workloads on the frontend on ROCm and AMD GPUs while maintaining compatibility with NVIDIA's cuDNN API."

Maybe:
"hipDNN allows developers to run deep learning workloads on AMD GPUs while maintaing compatibility with NVIDIA's cuDNN API."

Features
========

- **Graph-based API**: Operations are expressed as computational graphs rather than individual function calls, enabling optimization opportunities.
- **Plugin architecture**: Backend engines, heuristics, and benchmarking are implemented as plugins, allowing extensibility without modifying the core library.
- **Performance through Fusion**: Multiple operations can be fused into single kernels for better performance.
- **Engine selection**: Heuristics and benchmarking will be implemented as plugins, allowing extensibility without modifying the core library.

> - **Industry standard API**: Provides a familiar interface that matches established deep learning library conventions.

Reviewer: I changed this in the latest, I think this should be updated to match a bit:

https://github.com/ROCm/rocm-libraries/blob/develop/projects/hipdnn/docs/Design.md#core-design-principles

> 

Reviewer: I think we should have a section here talking about the components, and make it clear that the hipDNN frontend is how we expect users to consume hipDNN.

Design has a bit of a snippet on this:
```
Components
Frontend: A header-only C++ library that provides the industry standard API for interacting with hipDNN. The frontend wraps the backend C API to provide a more user-friendly C++ interface.

Backend: A shared library which provides a C API for hipDNN. The backend is the core component of hipDNN which acts as a plugin loader and manager, connecting problems to plugins that can solve them.

SDKs: Header-only libraries that provide shared utilities and interfaces. hipDNN provides three SDKs: Data SDK (graph schemas and data structures), Plugin SDK (plugin API and utilities), and Test SDK (testing utilities and CPU reference implementations).

Plugins: Plugins will be added over time to provide additional operational support, or performance improvements. Plugins should be external projects to hipDNN.
```

Maybe for the what-is section we can have something like:
```
Components
Frontend: A header-only C++ library that provides access to the hipDNN consumer API. hipDNN frontend supports describing kernel computational graphs, selecting and filtering backend engine providers, and executing the operation. 

Backend: A shared library which provides a C API for hipDNN to be used in conjunction with the hipDNN frontend. hipDNN backend acts as a plugin loader which is responsible for connecting computational graphs to the backend kernel providers that are capable of executing them. Note, it's recommended for users to access hipDNN through the frontend, and it's not required to interact directly with the backend.

SDKs: Header-only libraries that provide shared utilities and interfaces. hipDNN provides three SDKs: Data SDK (graph schemas and data structures), Plugin SDK (plugin API and utilities), and Test SDK (testing utilities and CPU reference implementations).

Plugins: Plugins will be added over time to provide additional operational support, or performance improvements. Plugins should be external projects to hipDNN.
```

The MIOpen legacy plugin serves as the kernel provider. It employs a modular C++ architecture, largely decoupled from the API layer.

- Dependency injection container (``MiopenContainer``): This is the root object that manages the lifecycle and dependencies of all other components. It initializes the ``EngineManager`` and ensures that all necessary services are correctly injected.
- Engine manager (``EngineManager``): The central registry for execution engines. It orchestrates the selection of the appropriate engine for a given operation graph by querying its registered engines.
- Plan builders (``IPlanBuilder``): Each engine is associated with a set of plan builders. These components are responsible for:

  - Applicability: Inspecting an operation graph to determine if the engine can execute it.
  - Resource estimation: Calculating the required workspace size.
  - Plan construction: Creating an executable IPlan object if the graph is supported.

- Plans (IPlan): An IPlan represents a strategy for executing a specific operation graph. It encapsulates all the necessary logic and state to run the routine, abstracting the details from the higher-level engine management.

> - C-API Interface: A thin translation layer that exposes these internal C++ components to the backend via the required engine plugin C-API.

Reviewer: Swap to MIOpen Provider Plugin, and I think might make sense to have a different page for each plugin project.

.. meta::

>    :description: hipDNN is a plugin-based deep learning library that provides graph-based operation support through backend plugins 

Reviewer 2: ```suggestion
   :description: hipDNN (Deep Neural Network) is a graph-based deep learning library that enables multi-operation fusion for improved performance on AMD GPUs. 
```

   :keywords: hipDNN, ROCm, library, API

.. _what-is:

******************
What is hipDNN?
******************

> hipDNN is a graph-based deep learning library that enables multi-operation fusion for improved performance on AMD GPUs. 

Reviewer 2: ```suggestion
hipDNN (Deep Neural Network) is a graph-based deep learning library that enables multi-operation fusion for improved performance on AMD GPUs.  
```

It uses operation graphs as an intermediate representation to describe computations, allowing different backend engines to optimize and execute these graphs efficiently.

> hipDNN allows developers to run deep learning workloads on AMD GPUs while maintaing compatibility with NVIDIA's cuDNN API.

Reviewer 2: ```suggestion
hipDNN allows developers to run deep learning workloads on AMD GPUs while maintaing compatibility with NVIDIA's cuDNN Frontend API.
```
Ref: https://docs.nvidia.com/deeplearning/cudnn/frontend/latest/index.html

Reviewer 3: Maybe we should be more precise with our wording here. "Compatibility" is a strong word given that any code would need to be hipified and there are additional API disparities (e.g. knobs, device properties).

Something like: "... _AMD GPUS by providing an interface modeled after NVIDIA's cuDNN Frontend API_."

hipDNN has a plugin-based architecture which allows advanced users to extend hipDNN without modifying the core library.

hipDNN and the ROCm ecosystem
=============================

.. list-table::
   :widths: 3 3 5
   :header-rows: 1

   * - Layer
     - Component
     - Function
   * - Framework
     - PyTorch/TensorFlow
     - The high-level AI tools developers use.
   * - Interface
     - **hipDNN**
     - The portable API for deep learning operations.
   * - Primitives
     - MIOpen
     - The AMD open-source, deep-learning primitives library for GPUs.
   * - Runtime
     - HIP
     - The translation layer between C++ and GPU hardware.
   * - Hardware
     - AMD Instinct / Radeon
     - The physical GPU.

Features
========

- **Graph-based API**: Operations are expressed as computational graphs rather than individual function calls, enabling optimization opportunities.

> - **Plugin architecture**: Backend kernel engines and heuristics are implemented through plugins, allowing extensibility without modifying the core library.

Reviewer 2: ```suggestion
- **Plugin architecture**: Backend kernel engines are implemented through plugins, allowing extensibility without modifying the core library.
```

- **Performance through fusion**: Multiple operations can be fused into single kernels for better performance.

> - **Engine selection**: Heuristics will be implemented as plugins, allowing extensibility without modifying the core library, and benchmarking will be implemented as an extensible frontend API allowing customized engine selection logic.

Reviewer 2: Heuristic plugins are not available in the initial release, so this point could be removed (unless it's preferred to keep this indicating something that hipDNN will have in a future release).

(@BrianHarrisonAMD to confirm?)

- **Industry standard API**: Provides a familiar interface that matches established deep learning library conventions.

Components
==========

- **Frontend**: A header-only C++ library that provides the industry standard API for interacting with hipDNN. The frontend wraps the backend C API to provide a more user-friendly C++ interface.

> - **Backend**: A shared library which provides a C API for hipDNN. The backend is the core component of hipDNN which acts as a plugin loader and manager, connecting problems to plugins that can solve them.

Reviewer 2: ```suggestion
- **Backend**: A shared library which provides a C API for hipDNN. The backend is the core component of hipDNN which acts as a plugin loader and manager, connecting problems to engines (provided via plugins) that can solve them.
```

- **SDKs**: Header-only libraries that provide shared utilities and interfaces. hipDNN provides three SDKs: Data SDK (graph schemas and data structures), Plugin SDK (plugin API and utilities), and Test SDK (testing utilities and CPU reference implementations).
- **Plugins**: Plugins provide additional operational support or performance improvements.

Plugins
=======

> The MIOpen Provider plugin serves as the kernel provider. It employs a modular C++ architecture, largely decoupled from the API layer. See :ref:`miopen` for more info.

Reviewer 2: To confirm: will the MIOpen provider plugin be the only plugin that is included in the initial hipDNN release?
